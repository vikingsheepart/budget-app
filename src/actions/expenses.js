import database from '../firebase/firebase';

export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  // set default values
  const {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0
  } = expenseData;

  return (dispatch, getState) => {
    const userId = getState().auth.uid;
    const expense = {
      description,
      note,
      amount,
      createdAt
    };

    // add expense to firebase and then dispatch action
    return database.ref(`users/${userId}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const userId = getState().auth.uid;
    return database.ref(`users/${userId}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }));
    });
  };
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const userId = getState().auth.uid;
    return database.ref(`users/${userId}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const userId = getState().auth.uid;
    return database.ref(`users/${userId}/expenses`).once('value')
      .then((snapshot) => {
        const expenses = [];
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
