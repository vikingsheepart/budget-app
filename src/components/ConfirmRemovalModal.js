import React from 'react';
import PropTypes from 'prop-types';
import Modal from 'react-modal';

Modal.defaultStyles.overlay.backgroundColor = 'translucent blue';

const ConfirmRemovalModal = props => (
  <Modal
    className="modal"
    contentLabel="Remove Expense"
    isOpen={props.isModalOpen}
    onRequestClose={props.onToggleModal} // esc or click outside to close modal
    closeTimeoutMS={200}
  >
    <h2 className="modal__title">Removing: {props.expense.description}</h2>
    <p className="modal__body">
      All the information will be deleted from the database.<br />
      This action cannot be undone.
    </p>
    <div className="modal__buttons">
      <button className="button" onClick={props.onToggleModal}>Keep Expense</button>
      <button className="button button--danger" onClick={props.onRemove}>Delete Expense</button>
    </div>
  </Modal>
);

ConfirmRemovalModal.propTypes = {
  expense: PropTypes.shape().isRequired,
  isModalOpen: PropTypes.bool.isRequired,
  onToggleModal: PropTypes.func.isRequired,
  onRemove: PropTypes.func.isRequired
};

export default ConfirmRemovalModal;
