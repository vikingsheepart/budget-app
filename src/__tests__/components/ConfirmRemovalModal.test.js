import React from 'react';
import { shallow } from 'enzyme';
import ConfirmRemovalModal from '../../components/ConfirmRemovalModal';
import expenses from '../fixtures/expenses';

let isModalOpen, onToggleModal, onRemove, wrapper;
beforeEach(() => {
  isModalOpen = true;
  onToggleModal = jest.fn();
  onRemove = jest.fn();
  wrapper = shallow(<ConfirmRemovalModal
    expense={expenses[2]}
    isModalOpen={isModalOpen}
    onToggleModal={onToggleModal}
    onRemove={onRemove}
  />);
});

test('should correctly render ConfirmRemovalModal', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onClick', () => {
  wrapper.find('button').at(0).simulate('click');
  wrapper.find('button').at(1).simulate('click');
  expect(onToggleModal).toHaveBeenCalled();
  expect(onRemove).toHaveBeenCalled();
});
