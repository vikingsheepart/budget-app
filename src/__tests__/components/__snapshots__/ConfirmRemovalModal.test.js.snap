// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ConfirmRemovalModal 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={200}
  contentLabel="Remove Expense"
  isOpen={true}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <h2
    className="modal__title"
  >
    Removing: 
    Credit Card
  </h2>
  <p
    className="modal__body"
  >
    All the information will be deleted from the database.
    <br />
    This action cannot be undone.
  </p>
  <div
    className="modal__buttons"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      Keep Expense
    </button>
    <button
      className="button button--danger"
      onClick={[Function]}
    >
      Delete Expense
    </button>
  </div>
</Modal>
`;
